### PR TITLE
Support for HTTP2 server push

### LINKED ISSUE #330: Add support for HTTP/2.0
See @dshafik issue at Guzzle for details. https://github.com/guzzle/guzzle/issues/1249

### PR DESCRIPTION
**Pull request by Nyholm**
This will fix #330 

There currently an issue where we do not get the response of the pushed responses. 

FYI @dshafik

### CODE REVIEW

File: lib/Browser.php [hunk 1]
```diff
@@ -268,6 +268,7 @@ private function prepareMultipart(string $name, string $content, string $boundar
     protected function createRequest(string $method, string $url, array $headers, $body): RequestInterface
     {
         $request = $this->requestFactory->createRequest($method, $url);
+        $request = $request->withProtocolVersion('2.0');
```
Comment (Nyholm) [inline lib/Browser.php]: This is just for debugging.

File: lib/Client/MultiCurl.php [hunk 2]
```diff
@@ -90,8 +96,23 @@ public function proceed(): void
...
+        if (!$this->curlm) {
+            if (false === $this->curlm = curl_multi_init()) {
+                throw new ClientException('Unable to create a new cURL multi handle');
+            }
+
+            $cb = function ($parent, $pushed, $headers) {
+                curl_setopt($pushed, CURLOPT_RETURNTRANSFER, true);
+                curl_setopt($pushed, CURLOPT_HEADER, false);
+                curl_setopt($pushed, CURLOPT_HEADERFUNCTION, null);
+                curl_setopt($pushed, CURLOPT_WRITEFUNCTION, null);
```
Comment (Nyholm) [inline lib/Client/MultiCurl.php]: These options do not seam to have much effect. 

Doing `curl_setopt($pushed, CURLOPT_WRITEFUNCTION, null);` does not help...

File: lib/Client/MultiCurl.php [hunk 3]
```diff
@@ -90,8 +96,23 @@ public function proceed(): void
...
+            }
+
+            $cb = function ($parent, $pushed, $headers) {
+                curl_setopt($pushed, CURLOPT_RETURNTRANSFER, true);
+                curl_setopt($pushed, CURLOPT_HEADER, true);
+                curl_setopt($pushed, CURLOPT_HEADERFUNCTION, null);
+                curl_setopt($pushed, CURLOPT_WRITEFUNCTION, null);
+                $this->addPushHandle($headers, $pushed);
+
+                return CURL_PUSH_OK;
```
Comment (Nyholm) [inline lib/Client/MultiCurl.php]: Make sure the caller can modify this somehow

File: lib/Client/MultiCurl.php [hunk 4]
```diff
@@ -90,8 +96,23 @@ public function proceed(): void
...
+        if (!$this->curlm) {
+            if (false === $this->curlm = curl_multi_init()) {
+                throw new ClientException('Unable to create a new cURL multi handle');
+            }
+
+            $cb = function ($parent, $pushed, $headers) {
+                curl_setopt($pushed, CURLOPT_RETURNTRANSFER, true);
+                curl_setopt($pushed, CURLOPT_HEADER, true);
+                curl_setopt($pushed, CURLOPT_HEADERFUNCTION, null);
+                curl_setopt($pushed, CURLOPT_WRITEFUNCTION, null);
```
Comment (Nyholm) [inline lib/Client/MultiCurl.php]: These options do not seam to have much effect.

Doing curl_setopt($pushed, CURLOPT_WRITEFUNCTION, null); does not help...

File: lib/Client/MultiCurl.php (hunk 3 above)
Comment (Nyholm) [reply lib/Client/MultiCurl.php]: Fixed

File: lib/Client/MultiCurl.php [hunk 5]
```diff
@@ -114,51 +174,128 @@ public function proceed(): void
...
+            }
+        }
+
+        if (!$found) {
+            $found = curl_getinfo($handle)['url'];
+        }
+
+        $content = curl_multi_getcontent($handle);
+        // Check if we got some headers, if not, we do not bother to store it.
+        if (0 !== $headerSize = curl_getinfo($handle, CURLINFO_HEADER_SIZE)) {
```
Comment (Nyholm) [inline lib/Client/MultiCurl.php]: This will never be true. I suspect a bug in PHP or Curl.

File: lib/Client/MultiCurl.php [hunk 6]
```diff
@@ -90,8 +116,31 @@ public function proceed(): void
...
+                foreach ($userCallbacks as $callback) {
+                    if (CURL_PUSH_DENY === $callback($parent, $pushed, $headers)) {
+                        return CURL_PUSH_DENY;
+                    }
+                }
+
+                curl_setopt($pushed, CURLOPT_RETURNTRANSFER, true);
+                curl_setopt($pushed, CURLOPT_HEADER, true);
+                curl_setopt($pushed, CURLOPT_HEADERFUNCTION, null);
+                curl_setopt($pushed, CURLOPT_WRITEFUNCTION, null);
```
Comment (Nyholm) [inline lib/Client/MultiCurl.php]: These options do not seam to have much effect.
Doing curl_setopt($pushed, CURLOPT_WRITEFUNCTION, null); does not help...

**Comment by Nyholm** [comment]
Blocked by https://bugs.php.net/bug.php?id=76675 and https://github.com/travis-ci/travis-ci/issues/9924

**Comment by nicolas-grekas** [comment]
Fix merged: https://github.com/php/php-src/pull/3744
Will require PHP >= 7.2.15 & libcurl >= 7.61.0

**Pull request merged**